crypto/mem.c

```diff
@@ -361,71 +198,145 @@ void *CRYPTO_realloc_clean(void *str, int old_len, int num, const char *file,
     if (str == NULL)
         return CRYPTO_malloc(num, file, line);
 
-    if (num <= 0)
+    if (num == 0) {
+        CRYPTO_clear_free(str, old_len);
         return NULL;
+    }
 
-    /*
-     * We don't support shrinking the buffer. Note the memcpy that copies
-     * |old_len| bytes to the new buffer, below.
-     */
-    if (num < old_len)
-        return NULL;
+    /* Can't shrink the buffer since memcpy below copies |old_len| bytes. */
+    if (num < old_len) {
+        memset((char*)str + num, 0, old_len - num);
+        return str;
+    }
+
+#ifdef CRYPTO_MDEBUG
+    if (call_malloc_debug) {
+        CRYPTO_mem_debug_realloc(str, NULL, num, file, line, 0);
+        ret = malloc(num);
+        CRYPTO_mem_debug_realloc(str, ret, num, file, line, 1);
+    } else {
+        ret = malloc(num);
+    }
+#else
+    (void)file;
+    (void)line;
+    ret = malloc(num);
+#endif
 
-    if (realloc_debug_func != NULL)
-        realloc_debug_func(str, NULL, num, file, line, 0);
-    ret = malloc_ex_func(num, file, line);
     if (ret) {
         memcpy(ret, str, old_len);
-        OPENSSL_clear_free(str, old_len);
+        CRYPTO_clear_free(str, old_len);
     }
-#ifdef LEVITTE_DEBUG_MEM
-    fprintf(stderr,
-            "LEVITTE_DEBUG_MEM:         | 0x%p -> 0x%p (%d)\n",
-            str, ret, num);
-#endif
-    if (realloc_debug_func != NULL)
-        realloc_debug_func(str, ret, num, file, line, 1);
 
     return ret;
 }
 
 void CRYPTO_free(void *str)
 {
-    if (free_debug_func != NULL)
-        free_debug_func(str, 0);
-#ifdef LEVITTE_DEBUG_MEM
-    fprintf(stderr, "LEVITTE_DEBUG_MEM:         < 0x%p\n", str);
+#ifdef CRYPTO_MDEBUG
+    if (call_malloc_debug) {
+        CRYPTO_mem_debug_free(str, 0);
+        free(str);
+        CRYPTO_mem_debug_free(NULL, 1);
+    } else {
+        free(str);
+    }
+#else
+    free(str);
 #endif
-    free_func(str);
-    if (free_debug_func != NULL)
-        free_debug_func(NULL, 1);
 }
 
 void CRYPTO_clear_free(void *str, size_t num)
 {
-    if (!str)
+    if (str == NULL)
         return;
     if (num)
         OPENSSL_cleanse(str, num);
     CRYPTO_free(str);
 }
 
-void *CRYPTO_remalloc(void *a, int num, const char *file, int line)
+char *CRYPTO_strdup(const char *str, const char* file, int line)
+{
+    char *ret;
+
+#ifndef CRYPTO_MDEBUG
+    (void)file;
+    (void)line;
+#endif
+    if (str == NULL)
+        return NULL;
+    ret = CRYPTO_malloc(strlen(str) + 1, file, line);
+    strcpy(ret, str);
```

Comment:
We could cache the strlen() output and use memcpy instead of strcpy, which might or might not be faster.  This version does not seem wrong, though.